### PR TITLE
Encoders for CSV

### PR DESCRIPTION
**Pull request by ybasket**

### CODE REVIEW

File: csv/generic/src/fs2/data/csv/generic/DerivedCellEncoder.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      cc: CellEncoder[H]): DerivedCellEncoder[A] =
+    a => cc.contramap[A](gen.to(_).head).apply(a)
+
+  // Coproducts
+
+  final implicit def coproductEncoder[T, Repr <: Coproduct](
+      implicit
+      gen: LabelledGeneric.Aux[T, Repr],
+      cc: Lazy[DerivedCellEncoder[Repr]]): DerivedCellEncoder[T] =
+    t => (cc.value: CellEncoder[Repr]).contramap(gen.to)(t)
```
Comment (satabin) [inline csv/generic/src/fs2/data/csv/generic/DerivedCellEncoder.scala:43]: Is the type ascription required?

File: csv/src/fs2/data/csv/CellEncoder.scala [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  implicit val booleanEncoder: CellEncoder[Boolean] = _.toString
+  implicit val byteEncoder: CellEncoder[Byte] = _.toString
+  implicit val shortEncoder: CellEncoder[Short] = _.toString
+  implicit val charEncoder: CellEncoder[Char] = _.toString
+  implicit val intEncoder: CellEncoder[Int] = _.toString
+  implicit val longEncoder: CellEncoder[Long] = _.toString
+  implicit val floatEncoder: CellEncoder[Float] = _.toString
+  implicit val doubleEncoder: CellEncoder[Double] = _.toString
+  implicit val bigDecimalEncoder: CellEncoder[BigDecimal] = _.toString
+  implicit val bigIntEncoder: CellEncoder[BigInt] = _.toString
```
Comment (satabin) [inline csv/src/fs2/data/csv/CellEncoder.scala:66]: I wonder whether it is not semantically more readable to use the `fromToString` function you introduced.

File: csv/src/fs2/data/csv/package.scala [hunk 3]
```diff
@@ -48,18 +52,64 @@ package object csv {
...
+
+  def encodeRow[F[_], Header, R](
+      implicit R: CsvRowEncoder[R, Header]): Pipe[F, R, CsvRow[Header]] =
+    _.map(R(_))
+
+  def encodeRowWithFirstHeaders[F[_], Header](
+      implicit H: WriteableHeader[Header])
+    : Pipe[F, CsvRow[Header], NonEmptyList[String]] =
+    _.pull.peek1.flatMap {
+      case Some((CsvRow(_, headers), tail)) =>
```
Comment (satabin) [inline csv/src/fs2/data/csv/package.scala:110]: `tail` is a really misleading name for `peek1`.

File: csv/src/fs2/data/csv/package.scala [hunk 4]
```diff
@@ -48,18 +52,64 @@ package object csv {
...
+          .map(RowWriter.encodeColumn(separator))
+          .intersperse(separator.toString) ++ Stream(newline))
+  }
+
+  def toRowStrings[F[_]](
+      separator: Char = ',',
+      newline: String = "\n"): Pipe[F, NonEmptyList[String], String] = {
+    // explicit Show avoids mapping the NEL before
+    val showColumn: Show[String] =
+      Show.show(RowWriter.encodeColumn(separator))
```
Comment (satabin) [inline csv/src/fs2/data/csv/package.scala:95]: I would make escaping optional, you might have a stream that is already escaped, or you know you don't need it.

Comment (ybasket) [reply csv/src/fs2/data/csv/package.scala]: Good point. I took this a little further and introduced an `EscapeMode` enum which allows to rely on automatic escaping, force escaping or completely turn it off.

File: csv/generic/src/fs2/data/csv/generic/DerivedCellEncoder.scala (hunk 1 above)
Comment (ybasket) [reply csv/generic/src/fs2/data/csv/generic/DerivedCellEncoder.scala]: Apparently not.

File: csv/src/fs2/data/csv/package.scala (hunk 3 above)
Comment (ybasket) [reply csv/src/fs2/data/csv/package.scala]: Oops…fixed.

File: csv/generic/src/fs2/data/csv/generic/DerivedCellEncoder.scala (hunk 1 above)
Comment (satabin) [reply csv/generic/src/fs2/data/csv/generic/DerivedCellEncoder.scala]: :sob:

**Pull request merged**